File: Notes.java
// Notes.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import com.google.gson.*;

class Note {
    String title;
    String content;
    List<Note> children;

    public Note(String title, String content) {
        this.title = title;
        this.content = content;
        this.children = new ArrayList<>();
    }

    public Note find(String title) {
        if (this.title.equals(title)) return this;
        for (Note child : children) {
            Note found = child.find(title);
            if (found != null) return found;
        }
        return null;
    }

    public Note findParent(String title, Note parent) {
        if (this.title.equals(title)) return parent;
        for (Note child : children) {
            Note found = child.findParent(title, this);
            if (found != null) return found;
        }
        return null;
    }

    public boolean delete(String title) {
        for (int i = 0; i < children.size(); i++) {
            if (children.get(i).title.equals(title)) {
                children.remove(i);
                return true;
            }
            if (children.get(i).delete(title)) return true;
        }
        return false;
    }

    public boolean move(String title, Note newParent) {
        Note note = find(title);
        if (note == null) return false;
        Note parent = findParent(title, null);
        if (parent == null) return false;
        parent.children.remove(note);
        newParent.children.add(note);
        return true;
    }

    public void display(int indent) {
        String prefix = children.isEmpty() ? "📄" : "📁";
        System.out.println("  ".repeat(indent) + prefix + " " + title);
        for (Note child : children) child.display(indent + 1);
    }

    public List<Note> search(String term) {
        List<Note> results = new ArrayList<>();
        if (title.toLowerCase().contains(term) || content.toLowerCase().contains(term))
            results.add(this);
        for (Note child : children)
            results.addAll(child.search(term));
        return results;
    }
}

public class Notes {
    private Note root;
    private final String filename = "notes.json";
    private Scanner scanner = new Scanner(System.in);
    private Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public Notes() {
        load();
    }

    public void load() {
        try {
            String json = new String(Files.readAllBytes(Paths.get(filename)));
            root = gson.fromJson(json, Note.class);
        } catch (Exception e) {
            root = new Note("Root", "");
        }
    }

    public void save(String filename) {
        if (filename == null) filename = this.filename;
        try {
            String json = gson.toJson(root);
            Files.write(Paths.get(filename), json.getBytes());
            System.out.println("Saved to " + filename);
        } catch (Exception e) {
            System.out.println("Save failed.");
        }
    }

    public void run() {
        System.out.println("🌳 Tree Notes");
        System.out.println("Commands: add, delete, move, rename, list, view, search, edit, export, import, save, quit");
        while (true) {
            System.out.print("> ");
            String line = scanner.nextLine().trim();
            String[] parts = line.split(" ");
            if (parts.length == 0) continue;
            String cmd = parts[0].toLowerCase();
            switch (cmd) {
                case "quit":
                    save(null);
                    System.out.println("Goodbye!");
                    return;
                case "save":
                    save(null);
                    break;
                case "add":
                    if (parts.length < 3) { System.out.println("Usage: add <parent> <title>"); break; }
                    String parentTitle = String.join(" ", Arrays.copyOfRange(parts, 1, parts.length - 1));
                    String title = parts[parts.length - 1];
                    Note parent = parentTitle.equals("root") ? root : root.find(parentTitle);
                    if (parent == null) { System.out.println("Parent '" + parentTitle + "' not found."); break; }
                    parent.children.add(new Note(title, ""));
                    System.out.println("Added note '" + title + "' under '" + parentTitle + "'");
                    break;
                case "delete":
                    if (parts.length != 2) { System.out.println("Usage: delete <title>"); break; }
                    if (root.delete(parts[1])) System.out.println("Deleted '" + parts[1] + "'");
                    else System.out.println("Note '" + parts[1] + "' not found.");
                    break;
                case "move":
                    if (parts.length < 3) { System.out.println("Usage: move <title> <new_parent>"); break; }
                    String moveTitle = parts[1];
                    String newParentTitle = String.join(" ", Arrays.copyOfRange(parts, 2, parts.length));
                    Note moveNote = root.find(moveTitle);
                    Note newParent = newParentTitle.equals("root") ? root : root.find(newParentTitle);
                    if (moveNote == null) System.out.println("Note '" + moveTitle + "' not found.");
                    else if (newParent == null) System.out.println("Parent '" + newParentTitle + "' not found.");
                    else if (root.move(moveTitle, newParent))
                        System.out.println("Moved '" + moveTitle + "' to '" + newParentTitle + "'");
                    break;
                case "rename":
                    if (parts.length != 3) { System.out.println("Usage: rename <old> <new>"); break; }
                    Note renameNote = root.find(parts[1]);
                    if (renameNote != null) { renameNote.title = parts[2]; System.out.println("Renamed '" + parts[1] + "' to '" + parts[2] + "'"); }
                    else System.out.println("Note '" + parts[1] + "' not found.");
                    break;
                case "list":
                    if (parts.length == 1) root.display(0);
                    else {
                        Note listNote = root.find(String.join(" ", Arrays.copyOfRange(parts, 1, parts.length)));
                        if (listNote != null) listNote.display(0);
                        else System.out.println("Note not found.");
                    }
                    break;
                case "view":
                    if (parts.length != 2) { System.out.println("Usage: view <title>"); break; }
                    Note viewNote = root.find(parts[1]);
                    if (viewNote != null) System.out.println("Title: " + viewNote.title + "\nContent: " + viewNote.content);
                    else System.out.println("Note not found.");
                    break;
                case "edit":
                    if (parts.length < 3) { System.out.println("Usage: edit <title> <content>"); break; }
                    Note editNote = root.find(parts[1]);
                    if (editNote != null) {
                        editNote.content = String.join(" ", Arrays.copyOfRange(parts, 2, parts.length));
                        System.out.println("Updated content for '" + parts[1] + "'");
                    } else System.out.println("Note not found.");
                    break;
                case "search":
                    if (parts.length != 2) { System.out.println("Usage: search <term>"); break; }
                    List<Note> results = root.search(parts[1].toLowerCase());
                    if (!results.isEmpty()) {
                        System.out.println("Found " + results.size() + " notes:");
                        for (Note r : results) {
                            Note p = root.findParent(r.title, null);
                            System.out.println("  " + r.title + " (parent: " + (p != null ? p.title : "root") + ")");
                        }
                    } else System.out.println("No results found.");
                    break;
                case "export":
                    if (parts.length != 2) { System.out.println("Usage: export <file>"); break; }
                    save(parts[1]);
                    break;
                case "import":
                    if (parts.length != 2) { System.out.println("Usage: import <file>"); break; }
                    try {
                        String json = new String(Files.readAllBytes(Paths.get(parts[1])));
                        root = gson.fromJson(json, Note.class);
                        System.out.println("Imported from " + parts[1]);
                    } catch (Exception e) { System.out.println("Failed to import."); }
                    break;
                default:
                    System.out.println("Unknown command.");
            }
        }
    }

    public static void main(String[] args) {
        new Notes().run();
    }
}

File: notes.js
// notes.js
const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

class Note {
    constructor(title, content = "") {
        this.title = title;
        this.content = content;
        this.children = [];
    }

    toJSON() {
        return {
            title: this.title,
            content: this.content,
            children: this.children.map(c => c.toJSON())
        };
    }

    static fromJSON(data) {
        const note = new Note(data.title, data.content || "");
        note.children = (data.children || []).map(c => Note.fromJSON(c));
        return note;
    }

    find(title) {
        if (this.title === title) return this;
        for (const child of this.children) {
            const found = child.find(title);
            if (found) return found;
        }
        return null;
    }

    findParent(title, parent = null) {
        if (this.title === title) return parent;
        for (const child of this.children) {
            const found = child.findParent(title, this);
            if (found !== null) return found;
        }
        return null;
    }

    delete(title) {
        for (let i = 0; i < this.children.length; i++) {
            if (this.children[i].title === title) {
                this.children.splice(i, 1);
                return true;
            }
            if (this.children[i].delete(title)) return true;
        }
        return false;
    }

    move(title, newParent) {
        const note = this.find(title);
        if (!note) return false;
        const parent = this.findParent(title);
        if (!parent) return false;
        const idx = parent.children.indexOf(note);
        if (idx > -1) parent.children.splice(idx, 1);
        newParent.children.push(note);
        return true;
    }

    display(indent = 0) {
        const prefix = this.children.length ? "📁" : "📄";
        console.log("  ".repeat(indent) + prefix + " " + this.title);
        this.children.forEach(c => c.display(indent + 1));
    }

    search(term) {
        let results = [];
        if (this.title.toLowerCase().includes(term) || this.content.toLowerCase().includes(term)) {
            results.push(this);
        }
        this.children.forEach(c => {
            results = results.concat(c.search(term));
        });
        return results;
    }
}

class App {
    constructor() {
        this.filename = "notes.json";
        this.root = new Note("Root");
        this.load();
    }

    load() {
        try {
            if (fs.existsSync(this.filename)) {
                const data = JSON.parse(fs.readFileSync(this.filename, 'utf8'));
                this.root = Note.fromJSON(data);
            }
        } catch (e) {
            this.root = new Note("Root");
        }
    }

    save(filename = this.filename) {
        fs.writeFileSync(filename, JSON.stringify(this.root.toJSON(), null, 2));
        console.log(`Saved to ${filename}`);
    }

    ask(question) {
        return new Promise(resolve => rl.question(question, resolve));
    }

    async run() {
        console.log("🌳 Tree Notes");
        console.log("Commands: add, delete, move, rename, list, view, search, edit, export, import, save, quit");
        while (true) {
            const input = await this.ask("> ");
            const parts = input.trim().split(/\s+/);
            if (!parts.length) continue;
            const cmd = parts[0].toLowerCase();
            switch (cmd) {
                case "quit":
                    this.save();
                    console.log("Goodbye!");
                    rl.close();
                    return;
                case "save":
                    this.save();
                    break;
                case "add":
                    if (parts.length < 3) {
                        console.log("Usage: add <parent> <title>");
                        break;
                    }
                    const parentTitle = parts.slice(1, -1).join(" ");
                    const title = parts[parts.length - 1];
                    const parent = parentTitle === "root" ? this.root : this.root.find(parentTitle);
                    if (!parent) { console.log(`Parent '${parentTitle}' not found.`); break; }
                    parent.children.push(new Note(title));
                    console.log(`Added note '${title}' under '${parentTitle}'`);
                    break;
                case "delete":
                    if (parts.length !== 2) { console.log("Usage: delete <title>"); break; }
                    const delTitle = parts[1];
                    if (this.root.delete(delTitle)) console.log(`Deleted '${delTitle}'`);
                    else console.log(`Note '${delTitle}' not found.`);
                    break;
                case "move":
                    if (parts.length < 3) { console.log("Usage: move <title> <new_parent>"); break; }
                    const moveTitle = parts[1];
                    const newParentTitle = parts.slice(2).join(" ");
                    const note = this.root.find(moveTitle);
                    const newParent = newParentTitle === "root" ? this.root : this.root.find(newParentTitle);
                    if (!note) console.log(`Note '${moveTitle}' not found.`);
                    else if (!newParent) console.log(`Parent '${newParentTitle}' not found.`);
                    else if (this.root.move(moveTitle, newParent)) {
                        console.log(`Moved '${moveTitle}' to '${newParentTitle}'`);
                    }
                    break;
                case "rename":
                    if (parts.length !== 3) { console.log("Usage: rename <old> <new>"); break; }
                    const oldTitle = parts[1], newTitle = parts[2];
                    const renameNote = this.root.find(oldTitle);
                    if (renameNote) { renameNote.title = newTitle; console.log(`Renamed '${oldTitle}' to '${newTitle}'`); }
                    else console.log(`Note '${oldTitle}' not found.`);
                    break;
                case "list":
                    if (parts.length === 1) this.root.display();
                    else {
                        const listTitle = parts.slice(1).join(" ");
                        const listNote = this.root.find(listTitle);
                        if (listNote) listNote.display();
                        else console.log("Note not found.");
                    }
                    break;
                case "view":
                    if (parts.length !== 2) { console.log("Usage: view <title>"); break; }
                    const viewNote = this.root.find(parts[1]);
                    if (viewNote) console.log(`Title: ${viewNote.title}\nContent: ${viewNote.content}`);
                    else console.log("Note not found.");
                    break;
                case "edit":
                    if (parts.length < 3) { console.log("Usage: edit <title> <content>"); break; }
                    const editTitle = parts[1];
                    const content = parts.slice(2).join(" ");
                    const editNote = this.root.find(editTitle);
                    if (editNote) { editNote.content = content; console.log(`Updated content for '${editTitle}'`); }
                    else console.log("Note not found.");
                    break;
                case "search":
                    if (parts.length !== 2) { console.log("Usage: search <term>"); break; }
                    const results = this.root.search(parts[1].toLowerCase());
                    if (results.length) {
                        console.log(`Found ${results.length} notes:`);
                        results.forEach(r => {
                            const parent = this.root.findParent(r.title);
                            const pTitle = parent ? parent.title : "root";
                            console.log(`  ${r.title} (parent: ${pTitle})`);
                        });
                    } else console.log("No results found.");
                    break;
                case "export":
                    if (parts.length !== 2) { console.log("Usage: export <file>"); break; }
                    this.save(parts[1]);
                    break;
                case "import":
                    if (parts.length !== 2) { console.log("Usage: import <file>"); break; }
                    try {
                        const data = JSON.parse(fs.readFileSync(parts[1], 'utf8'));
                        this.root = Note.fromJSON(data);
                        console.log(`Imported from ${parts[1]}`);
                    } catch (e) { console.log("Failed to import."); }
                    break;
                default:
                    console.log("Unknown command.");
            }
        }
    }
}

const app = new App();
app.run().catch(console.error);
